### PR TITLE
Update wasm related dependencies

### PR DESCRIPTION
**Pull request by benbrandt**
This updates all of the WASM related dependencies I could.

I think there is one failing test still, which I don't understand, so it is possible I messed something up in addressing the changes. Would appreciate a lookover.

Also, I couldn't go to the latest for everything because some of the sub git deps are pinned at wasm-tools 0.209.1, so I had to keep them there as well as a wasmtime version that also used those versions.

Hopefully this is a helpful start though, and if not, no worries :)

### CODE REVIEW

File: src/bindings.rs [hunk 1]
```diff
@@ -112,6 +112,7 @@ pub fn make_bindings(
         EntityType::Global(GlobalType {
             val_type: ValType::I32,
             mutable: false,
+            shared: false,
```
Comment (benbrandt) [inline src/bindings.rs:115]: The docs all had this as `false`, but I don't know what the correct value should be.

File: src/bindings.rs [hunk 2]
```diff
@@ -135,17 +137,15 @@ pub fn make_bindings(
             maximum: None,
             memory64: false,
             shared: false,
+            page_size_log2: None,
```
Comment (benbrandt) [inline src/bindings.rs:140]: Same here. I assumed the default

File: src/bindings.rs [hunk 3]
```diff
@@ -135,17 +137,15 @@ pub fn make_bindings(
...
 
     imports.import(
         "env",
         "__indirect_function_table",
         EntityType::Table(TableType {
-            element_type: RefType {
-                nullable: true,
-                heap_type: HeapType::Func,
-            },
+            element_type: RefType::FUNCREF,
```
Comment (benbrandt) [inline src/bindings.rs:148]: It seemed that this constant was a convenience wrapper around the same thing, but I could be wrong

File: src/bindings.rs [hunk 4]
```diff
@@ -154,6 +154,7 @@ pub fn make_bindings(
                 .try_into()
                 .unwrap(),
             maximum: None,
+            table64: false,
```
Comment (benbrandt) [inline src/bindings.rs:157]: Followed the doc example here as well

File: src/lib.rs [hunk 5]
```diff
@@ -562,30 +558,12 @@ pub async fn componentize(
...
-        )
-        .preopened_dir(
-            Dir::open_ambient_dir(bundled.path(), cap_std::ambient_authority())
-                .with_context(|| format!("unable to open {}", bundled.path().display()))?,
-            DirPerms::all(),
-            FilePerms::all(),
-            "bundled",
-        );
+        .preopened_dir(stdlib.path(), "python", DirPerms::all(), FilePerms::all())?
+        .preopened_dir(bundled.path(), "bundled", DirPerms::all(), FilePerms::all())?;
```
Comment (benbrandt) [inline src/lib.rs:562]: It seemed that this now just wanted paths. I think I got the arguments in the right order...

File: src/lib.rs [hunk 6]
```diff
@@ -787,10 +759,9 @@ fn parse_wit(path: &Path, world: Option<&str>) -> Result<(Resolve, WorldId)> {
 fn add_wasi_and_stubs(
     resolve: &Resolve,
     worlds: &IndexSet<WorldId>,
-    component: &Component,
     linker: &mut Linker<Ctx>,
 ) -> Result<()> {
-    wasi_command::add_to_linker(linker)?;
+    wasmtime_wasi::add_to_linker_async(linker)?;
```
Comment (benbrandt) [inline src/lib.rs:764]: I was having trouble finding if the old `command` was just replaced by the top-level linker. it is possible this is linking too much, I couldn't tell.

File: src/test/echoes.rs [hunk 7]
```diff
@@ -8,13 +8,13 @@ use {
...
     },
-    wasmtime_wasi::preview2::command,
 };
 
 wasmtime::component::bindgen!({
     path: "src/test/wit",
     world: "echoes-test",
-    async: true
+    async: true,
+    trappable_imports: true,
```
Comment (benbrandt) [inline src/test/echoes.rs:17]: I added this to all of them, since the previous implementations allowed for `Result` in the return values.

File: src/test/tests.rs [hunk 8]
```diff
@@ -8,26 +8,22 @@ use {
...
     with: {
         "componentize-py:test/resource-import-and-export/thing": ThingU32,
         "componentize-py:test/resource-borrow-import/thing": ThingU32,
-        "componentize-py:test/resource-borrow-export/thing": ThingU32,
         "componentize-py:test/resource-with-lists/thing": ThingList,
         "componentize-py:test/resource-aggregates/thing": ThingU32,
         "componentize-py:test/resource-alias1/thing": ThingString,
         "componentize-py:test/resource-floats/float": MyFloat,
         "resource-floats-imports/float": MyFloat,
-        "resource-floats-exports/float": MyFloat,
```
Comment (benbrandt) [inline src/test/tests.rs:30]: These came up as "unused"

**Review by benbrandt — COMMENTED** [review]
I didn't touch pyo3 because that might be a bigger change that I didn't get to.
I also was having issues in the `runtime` crate updating `wit-bindgen`, but I'm not sure if that needs to be updated or not

**Comment by benbrandt** [comment]
@dicej I added some comments where I was unsure on things to help focus the review. I don't claim to be an expert, and the changelogs are hard to follow/nonexistent for these crates

**Comment by dicej** [comment]
Thanks, @benbrandt !  BTW, you probably noticed I was using forks of `wasmtime` etc. with patches to support [isyswasfa](https://github.com/dicej/isyswasfa).  We can drop all that and remove the `isyswasfa` code -- it was a proof-of-concept that has served its purpose; we don't need it anymore.

BTW, I'd be fine with ripping off the band-aid and upgrading to `wasmtime` 24 and whatever `wasm-tools` it's using.  Might as well jump to the latest of everything while we're doing this.

**Comment by benbrandt** [comment]
> Thanks, @benbrandt ! BTW, you probably noticed I was using forks of `wasmtime` etc. with patches to support [isyswasfa](https://github.com/dicej/isyswasfa). We can drop all that and remove the `isyswasfa` code -- it was a proof-of-concept that has served its purpose; we don't need it anymore.
> 
> BTW, I'd be fine with ripping off the band-aid and upgrading to `wasmtime` 24 and whatever `wasm-tools` it's using. Might as well jump to the latest of everything while we're doing this.

Ok I can take a look at it tomorrow. I created a few more PRs for downstream deps so that the other stuff can be updated.

https://github.com/dicej/wasm-convert/pull/1/files
https://github.com/dicej/component-init/pull/2
https://github.com/dicej/isyswasfa-transform/pull/1

Though the last one isn't necessary if we rip it out.

**Comment by benbrandt** [comment]
@dicej I removed the isyswasfa, but I am really struggling with some of the wasmtime changes, especially in the test generators. I'll see if I can manage to get this in an ok state

**Comment by benbrandt** [comment]
@dicej ok I pushed up what I did. It compiles, but tests fail horribly. Maybe you might have some ideas of what I did wrong here 🙈

File: src/lib.rs [hunk 9]
```diff
@@ -125,9 +122,7 @@ impl Invoker for MyInvoker {
     async fn call_s32(&mut self, function: &str) -> Result<i32> {
         let func = self
             .instance
-            .exports(&mut self.store)
-            .root()
-            .typed_func::<(), (i32,)>(function)?;
+            .get_typed_func::<(), (i32,)>(&mut self.store, function)?;
```
Comment (benbrandt) [inline src/lib.rs:125]: These all seem to fail, which I suspect is because somehow I am not retrieving the function from the correct level.

File: src/lib.rs [hunk 10]
```diff
@@ -742,15 +683,18 @@ pub async fn componentize(
...
                 if !added_to_linker {
-                    add_wasi_and_stubs(&resolve, &worlds, component, &mut linker)?;
+                    add_wasi_and_stubs(&resolve, &worlds, &mut linker)?;
                 }
 
-                let (init, instance) =
-                    Init::instantiate_async(&mut store, component, &linker).await?;
+                let instance = linker.instantiate_async(&mut store, component).await?;
+                let init = instance
+                    .get_typed_func::<(&str, &Symbols, bool), (Result<(), String>,)>(
```
Comment (benbrandt) [inline src/lib.rs:691]: I had to replicate some of what was happening in the bindgen because the bindgen code no longer returns the instance

File: src/test.rs [hunk 11]
```diff
@@ -106,10 +105,7 @@ trait Host {
 
     fn add_to_linker(linker: &mut Linker<Ctx>) -> Result<()>;
 
-    async fn instantiate_pre(
-        store: &mut Store<Ctx>,
-        pre: &InstancePre<Ctx>,
-    ) -> Result<(Self::World, Instance)>;
+    async fn instantiate_pre(store: &mut Store<Ctx>, pre: InstancePre<Ctx>) -> Result<Self::World>;
```
Comment (benbrandt) [inline src/test.rs:108]: Since the instance doesn't get returned anymore, I removed it from this return as well.
Also, the only methods I could find that did roughly the same thing needed to consume the InstancePre. Since this is holding Arcs, I am hoping it is ok to clone and consume...

File: test-generator/src/lib.rs [hunk 12]
```diff
@@ -699,7 +699,7 @@ pub fn generate() -> Result<()> {
 
             writeln!(
                 &mut typed_function_inits,
-                r#"echo{test_index}: instance.typed_func::<({params}), ({result_type},)>("echo{test_index}")?,"#
+                r#"echo{test_index}: instance.get_typed_func::<({params}), ({result_type},)>(&mut *store, "componentize-py:test/echoes-generated/echo{test_index}")?,"#
```
Comment (benbrandt) [inline test-generator/src/lib.rs:702]: I'm almost positive this isn't correct....

File: test-generator/src/lib.rs [hunk 13]
```diff
@@ -831,21 +831,19 @@ impl super::Host for Host {{
...
-        let guest_instance = pre.instantiate_async(&mut *store).await?;
-        let mut exports = guest_instance.exports(&mut *store);
-        let mut instance = exports.instance("componentize-py:test/echoes-generated").unwrap();
+        pre: InstancePre<Ctx>,
+    ) -> Result<Self::World> {{
+        let instance = pre.instantiate_async(&mut *store).await?;
         Ok((Self::World {{
            {typed_function_inits}
-        }}, guest_instance))
+        }}))
```
Comment (benbrandt) [inline test-generator/src/lib.rs:846]: This too. All of this has changed a lot, and I don't think I am accessing the functions correctly here.

**Comment by dicej** [comment]
@benbrandt Thanks for putting so much effort into this.  I'm having a busy week, so apologies for not helping much so far.  Hopefully I'll have some time tomorrow to take a look.

**Comment by benbrandt** [comment]
No worries. If you want you can also just go commit by commit and if I went wrong somewhere I can maybe cherry pick or revert if the others are ok

**Comment by dicej** [comment]
@benbrandt I just pushed an update to fix the tests.  I'll merge your PRs on the other repos, then update the Cargo.toml here to match, and then we should be ready to merge this.

File: src/bindings.rs (hunk 2 above)
Comment (dicej) [reply src/bindings.rs:140]: Yeah, we have no use for non-default page sizes at this point.  That's primarily meant for embedded systems.

**Comment by dicej** [comment]
@benbrandt Thanks again for doing this; it saved me a _lot_ of time!

**Comment by benbrandt** [comment]
@dicej glad I could help!

**Comment by dicej** [comment]
FYI, I'm planning to publish a new release to pypi.org, but having CI issues I may need an administrator to help with, and that may need to wait until Monday.

**Comment by dicej** [comment]
Update: the release is live: https://pypi.org/project/componentize-py/

**Comment by benbrandt** [comment]
Amazing thank you!
